### ISSUE TITLE
Dependency cycle detected  import/no-cycle

### ISSUE THREAD

**Issue by tcastelly** [benmosher/eslint-plugin-import#2061]
Hello,

After an update from '2.22.1' to '2.23.0' I have this error
>  Dependency cycle detected  import/no-cycle

I'm trying to import a TypeScript Type like this:

./Bar.ts
```javascript
import useBar from './useBar';

export type Bar = {
  msg: string
}
```
---
./useBar.ts
```javascript
import type { Bar } from './Bar';
```

Previously import type even if it was cycled import worked.

**Comment by ljharb** [comment]
That is indeed a dependency cycle, so it's a bug that it wasn't warning on it before.

**Comment by tcastelly** [comment]
Was really a bug before?
According to this issue, https://github.com/benmosher/eslint-plugin-import/issues/1453 I thought because we can import type with 
```javascript
import type from { ... }
```

it was allowed.

Anyway, thank you for your answer.

**Comment by ljharb** [comment]
No, anything that causes two files to depend on each other is a cycle, and is a horrifically bad architecture, which is the purpose of the rule to prevent.

**Comment by raycharius** [comment]
Sorry, @ljharb, but this is wrong. The purpose of this rule is to avoid possible issues at *runtime* with circular dependencies. Using `import type` will *never* lead to that. Moreover, I disagree about this always being bad architecture when it comes to importing types. There are plenty of cases when this is more than appropriate.

At the very least, it should be configurable to exclude `import type`.  

The makers of [Madge](https://github.com/pahen/madge#how-to-ignore-import-type-statements-in-es6--flow) know and understand that.

In any case, this has rendered this check useless for me because of this, which is actually really unfortunate. So, turning this rule off and integrating Madge into CI flow.

**Comment by ljharb** [comment]
I agree that there’s a difference between avoiding runtime and build time cycles, and I’d be fine adding an option that allows type cycles.

However, `import type` can absolutely lead to that because enums and classes are in both value and type space.

**Comment by gabts** [comment]
I agree a flag would be nice, TypeScript already yells at me if I try to use imported type as value.

![Screenshot 2021-05-17 at 16 00 00](https://user-images.githubusercontent.com/31779292/118501573-28bdc680-b729-11eb-9d2d-11c021e1ebed.png)

**Comment by quvide** [comment]
As @raycharius commented, this definitely is a breaking change for many projects since the rule now flags a [recommended pattern in Redux](https://redux.js.org/recipes/usage-with-typescript#define-slice-state-and-action-types) where the root state type is inferred by TypeScript from imported reducers.

> You can safely import the RootState type from the store file here. It's a circular import, but the TypeScript compiler can correctly handle that for types. This may be needed for use cases like writing selector functions.
